stream: improve performance of finished()

avcribl opened the pull request:
The changes optimize the `end-of-stream.js` module to only preserve AsyncLocalStorage context when it's actually needed, improving performance by avoiding ALS overhead in the common case where no async context tracking is active.
The performance test results with this patch:
```
./out/Release/node ./benchmark/streams/end-of-stream.js
streams/end-of-stream.js streamType="readable" n=100000: 3,290,307.950450068
streams/end-of-stream.js streamType="writable" n=100000: 1,004,954.4253168119
streams/end-of-stream.js streamType="duplex" n=100000: 893,658.706842607
```
The same perf test results with the current nodejs:
```
node ./benchmark/streams/end-of-stream.js
streams/end-of-stream.js streamType="readable" n=100000: 385,512.31905964843
streams/end-of-stream.js streamType="writable" n=100000: 348,685.1671624238
streams/end-of-stream.js streamType="duplex" n=100000: 382,477.79857896036
```
You can compare and see the noticeable performance improvement with this patch.

mcollina reviewed (changes requested):
While I love the PR, but I fear that this breaks continuation when using async_hooks.

We could argue that with ALS not using async_hooks we could remove them (or add them via a compile time flag) but that's outside of the scope of this PR.

avcribl commented on benchmark/streams/end-of-stream.js:
```diff
@@ -0,0 +1,38 @@
+    switch (streamType) {
+      case 'readable':
+        stream = new Readable({ read() { this.push(null); } });
+        break;
```
Thanks, added!

avcribl commented on benchmark/streams/end-of-stream.js:
```diff
@@ -0,0 +1,38 @@
+        break;
+    }
+
+    finished(stream, () => {});
```
Thanks, replaced it with await one.

avcribl commented on benchmark/streams/end-of-stream.js:
```diff
@@ -0,0 +1,38 @@
+
+const common = require('../common');
+const { Readable, Writable, Duplex, finished } = require('stream');
+
```
Thanks, done.

avcribl commented:
> While I love the PR, but I fear that this breaks continuation when using async_hooks.
> 
> We could argue that with ALS not using async_hooks we could remove them (or add them via a compile time flag) but that's outside of the scope of this PR.

Thanks @mcollina! I removed the async_hooks part.

krsjenswbp reviewed (commented):
thanks

Flarna commented:
There is command line option `--no-async-context-frame` to disable use of `AsyncContextFrame` in ALS.

I think we might need special handing for this variation.

avcribl commented:
> @avcribl you didn't address my comment.

@mcollina I've pushed a new commit that adds a check to determine whether async_context_frame is enabled. I tested both scenarios:
when async_context_frame is enabled, and when it is disabled (falling back to async_hooks).

In both cases, the unit tests ran successfully and passed:
```
./out/Release/node --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /Users/amir/Documents/projects4/node/test/async-hooks/test-async-local-storage-stream-finished.js

./out/Release/node --no-async-context-frame --test-reporter=./test/common/test-error-reporter.js --test-reporter-destination=stdout /Users/amir/Documents/projects4/node/test/async-hooks/test-async-local-storage-stream-finished.js
```
Also, I ran the benchmark test with and without the flag and both have similar improved performance:
```
./out/Release/node ./benchmark/streams/finished.js
streams/finished.js streamType="readable" n=100000: 1,033,352.7517537934
streams/finished.js streamType="writable" n=100000: 1,383,377.7808003046

./out/Release/node --no-async-context-frame ./benchmark/streams/finished.js
streams/finished.js streamType="readable" n=100000: 1,044,042.0363062695
streams/finished.js streamType="writable" n=100000: 1,362,108.6226052495
```
Please let me know if it addresses your comment.

Flarna commented:
It's quite hard to tell. I would recommend to write tests for all these scenarios to be improved here and also verify that non improvable scenarios still work.

FWIW I'm not a big fan of tinkering with async hooks/async local store internals at several places in code.
Maybe using `enabledHooksExist` instead `getHookArrays()[0].length` would make it easier to understand.

Qard commented:
One thing worth noting is that this would be a breaking change as currently the AsyncLocalStorage.bind(...) will _always_ bind, even if there are no hooks or stores in use, but async_hooks could start listening at any time. In the current model you could start listening _after_ the bind call there but _before_ the callback is called and you would see the events. After this change you would not.

avcribl commented on lib/internal/streams/end-of-stream.js:
```diff
@@ -66,8 +69,13 @@ function eos(stream, options, callback) {
-  AsyncLocalStorage ??= require('async_hooks').AsyncLocalStorage;
-  callback = once(AsyncLocalStorage.bind(callback));
+  if ((AsyncContextFrame.enabled && AsyncContextFrame.current()) ||
+      (!AsyncContextFrame.enabled && getHookArrays()[0].length > 0)) {
```
Thanks! I removed `AsyncContextFrame.current()`

Flarna commented on test/parallel/test-stream-finished-bindAsyncResource-path.js:
```diff
@@ -0,0 +1,21 @@
+finished(readable, common.mustCall(() => {
+  strictEqual(internalAsyncHooks.getHookArrays()[0].length > 0,
+              true, 'Should have active user async hook');
+}));
```
This doesn't verify the code path either. It is a copy of the condition and someone might change one of the locations.

avcribl replied:
Thanks! I expanded on the test!

avcribl replied:
@mcollina I addressed your comment, thanks!

mcollina commented:
cc @Flarna

mcollina commented:
@Qard are you still sure this is a major?

Qard commented:
Well, I suppose async_hooks is _technically_ still marked experimental, so we _can_ consider it non-breaking. It definitely _does_ break async_hooks slightly though as this will change the model to not trigger later events if the init was missed. With AsyncLocalStorage not using it anymore, I don't really care so much for _that_ use case, but others still _using_ async_hooks _might_ care. Though, in my opinion, other events triggering despite earlier ones being missed has been as much a footgun for people, so... 🤷🏻

avcribl commented:
@mcollina Is there any update for my PR? Thanks!

avcribl commented:
Can the pipeline be triggered again? Still, some flaky tests fail, thank you!

avcribl commented:
Thanks @H4ad ! I think the label `commit-queue` needs to be added as well.
